msu/systems/serialization/serialization_types/custom_serialization_data.nut

```diff
@@ -0,0 +1,147 @@
+::MSU.Class.CustomSerializationData <- class extends ::MSU.Class.AbstractSerializationData
+{
+	__MetaData = null;
+
+	constructor(_data)
+	{
+		base.constructor(_data);
+	}
+
+	// must be called when overriden
+	function deserialize( _in )
+	{
+		local type = _in.readU8();
+		if (type != this.DataType.U32)
+			throw ::MSU.Exception.InvalidValue(type);
+		this.setLength(_in.readU32());
+	}
+
+	// must be called when overriden
+	function serialize( _out )
+	{
+		base.serialize(_out);
+		::MSU.Class.U32SerializationData(this.len()).serialize(_out); // store length
+	}
+
+	function setLength( _length )
+	{
+		throw "TODO" // must be implemented for all containers inheriting from this object
+	}
+
+	function len()
+	{
+		throw "TODO" // must be implemented for all containers inheriting from this object
+	}
+
+	function __readValueFromStorage( _in )
+	{
+		local type = _in.readU8();
+		switch (type)
+		{
+			case this.DataType.Null:
+				return ::MSU.Class.NullSerializationData();
+			case this.DataType.Bool:
+				return ::MSU.Class.BoolSerializationData(_in.readBool());
+			case this.DataType.String:
+				return ::MSU.Class.StringSerializationData(_in.readString());
+			case this.DataType.U8:
+				return ::MSU.Class.U8SerializationData(_in.readU8());
+			case this.DataType.U16:
+				return ::MSU.Class.U16SerializationData(_in.readU16());
+			case this.DataType.U32:
+				return ::MSU.Class.U32SerializationData(_in.readU32());
+			case this.DataType.I8:
+				return ::MSU.Class.I8SerializationData(_in.readI8());
+			case this.DataType.I16:
+				return ::MSU.Class.I16SerializationData(_in.readI16());
+			case this.DataType.I32:
+				return ::MSU.Class.I32SerializationData(_in.readI32());
+			case this.DataType.F32:
+				return ::MSU.Class.F32SerializationData(_in.readF32());
+			case this.DataType.DataArray:
+				local dataArray = ::MSU.Class.DataArrayData();
+				dataArray.deserialize(_in);
+				return dataArray;
+			case this.DataType.Table:
+				local table = ::MSU.Class.TableSerializationData(null);
```

Comment:
Ah here you pass null so that's why you need `null` handling in the constructors. Instead of that I'd just pass an empty table or array here as needed and remove the null handling from the constructor in those classes.